feldera/feldera
# pipeline-manager: compiler server improvements including caching

Addresses the following in the compiler server:
- Rust compilation now uses caching, such that a program with the same Rust code will not be recompiled but instead the already existing binary is used. Whether a program already exists in the cache is determined by a source checksum which is calculated over all the inputs that should influence it, including platform version.
- Decouple SQL and Rust compilation: when there is an ongoing Rust compilation (which can take quite some time), SQL compilation (i.e., validation) continues to work for other pipelines.
- More thorough cleanup of SQL and Rust compilation artifacts, which prevents the unnecessary buildup of disk usage over time.
- Versioning of the platform through the use of `platform_version`. Legacy pipelines that are already running will continue to run. Starting a legacy pipeline anew will trigger an automatic upgrade and recompilation, after which it will start. This is through the use of the "early start" mechanism, in which a pipeline can already be issued the `/start` irrespective of the program status. The pipeline deployment will becomes `Failed` if the program fails or has failed to compile.

The working directory is restructured to separate the compiler working directory from that of the local-runner. Within the compiler working directory, SQL and Rust compilation are also separated, as they are now decoupled. This significantly simplifies working directory maintenance, as it is clear where which state is stored and used by what.

**Related issues:**
- Fixes: https://github.com/feldera/feldera/issues/1741

**Next PR:**
- Field filters for API server endpoints GET pipeline(s) endpoints to reduce size (new `main_rust` in `program_info` can be large)
- Additional tests beyond the database tests

**Future work PRs:**
- Retain the cache longer instead of deleting it immediately in the next cleanup cycle
- Regularly report disk space usage in the logs
- Add a `/recompile` endpoint to trigger recompilation manually rather than needing a trivial edit
- Improve handling of `Cargo.lock`
- Allow passing SQL compiler flags
- Support for multiple compiler servers

**Steps to test the automatic upgrade:**
1. Run manager: `cargo run --package=pipeline-manager --features pg-embed --bin pipeline-manager -- --dev-mode`
2. Create one of the demos, wait for it to compile and start it
3. Terminate the manager from step (1)
4. Run manager with a new platform version: `cargo run --package=pipeline-manager --features pg-embed --bin pipeline-manager -- --dev-mode --platform-version v2`
5. Start the demo again (you'll have to shut it down if you have not before), it should now compile again and then afterwards will automatically start
6. The manager logs are informative to what is happening

**Remaining tasks for this PR:**
- [x] Upgrading based on `platform_version` (@snkas)
- [x] Implicit upgrade when issuing `/start` (@snkas)
- [x] UI: handle `SqlCompiled` and program configuration JSON dialog (@Karakatiza666)
- [x] fda: handle `SqlCompiled` and program configuration `cache` (@snkas) -- It does not yet support providing a value for `cache`, instead just does not provide it which means it uses default (`true`).
- [x] python: handle `SqlCompiled` and program configuration `cache` (@snkas) -- It does not yet support providing a value for `cache`, instead just does not provide it which means it uses default (`true`).
